### DISCUSSION TITLE
How can I retrieve HTTP response headers in Tauri 2?

### DISCUSSION THREAD

**Discussion by xrgfn153** [tauri-apps/tauri#11983] [Q&A]
I'm using the JS API, and after upgrading to version 2.0, the headers I'm receiving are empty. I've tried configuring tauri-plugin-http = { version = "2", features = ["unsafe-headers"] } as per the documentation, but it hasn't worked.

**Comment by bryantamayo1** [comment]
Hi @xrgfn153!
In my case I have the following configuration (project is developed on Ubuntu 22.04.4 LTS ):
<br>
_src-tauri/Cargo.toml_
```
[dependencies]
...
tauri = { version = "2.4.1", features = []  }
tauri-plugin-http = { version = "2", features = ["unsafe-headers"] }
```

> More information about "unsafe-headers" in [http-client plugin](https://v2.tauri.app/plugin/http-client/)

<br>

_src-tauri/capabilities/default.json_
```
{
...,
  "permissions": [
    "core:default",
    "http:allow-fetch",
    "http:allow-fetch-cancel",
    "http:allow-fetch-read-body",
    "http:allow-fetch-send",
    {
      "identifier": "http:default",
      "allow": [
        {
          "url": "https://*"
        }
      ]
    }
  ]
}
```

>If you need more information about permissions, this path can help you _src-tauri/gen/schemas/desktop-schema.json_
<br>

_src-tauri/src/lib.rs_

```
  ...
  tauri::Builder::default()
  // Add this plugin here
  .plugin(tauri_plugin_http::init())
  .setup(|app| {
    if cfg!(debug_assertions) {
      app.handle().plugin(
        tauri_plugin_log::Builder::default()
          .level(log::LevelFilter::Info)
          .build(),
      )?;
    }
    Ok(())
  })
  .run(tauri::generate_context!())
  .expect("error while running tauri application");
  ...
```
<br>

> In your web application, use this:
```
  ...
  import { fetch} from '@tauri-apps/plugin-http';
  const optionsFetch: RequestInit = {
    method: 'GET',
    responseType: 'Json',
  };
  // More about types in: node_modules/@tauri-apps/plugin-http/dist-js/index.d.ts
  const resp = await fetch('http://your-url', optionsFetch);
  const data = await resp.json();
  ...
```
<br>
My web application has a server with https and nginx, for these reasons it works well. My goal is to achieve the web application works with https and self-signed certificate but unluckily I don't get it 😟. I hope these tips can help you, to be honest, it is really difficult to find information about Tauri v2 since it launched on Octuber 2024.

**Reply by rudolphi** [reply]
When I test, resp.headers["set-cookie"] and resp.headers.location are not populated (in fact the map is empty)

**Comment by Think-M** [comment]
@FabianLars Do you have any solutions? : )

**Reply by FabianLars** [reply]
I'm not aware of any issues. With the latest plugin versions i can see the response headers just fine.
![grafik](https://github.com/user-attachments/assets/4336379c-7ba6-421f-9615-a3fe2ca60c8c)

The response.headers object is of type https://developer.mozilla.org/en-US/docs/Web/API/Headers so you i think you have to iterate over it to get the headers.

If it's not that, can you share some code that isn't working and the output of `tauri info` ?

**Reply by Think-M** [reply]
> I'm not aware of any issues. With the latest plugin versions i can see the response headers just fine. ![grafik](https://private-user-images.githubusercontent.com/30730186/452923944-4336379c-7ba6-421f-9615-a3fe2ca60c8c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDk2MDI4MjcsIm5iZiI6MTc0OTYwMjUyNywicGF0aCI6Ii8zMDczMDE4Ni80NTI5MjM5NDQtNDMzNjM3OWMtN2JhNi00MjFmLTk2MTUtYTNmZTJjYTYwYzhjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA2MTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNjExVDAwNDIwN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTEyMDFmMTFlMGRlMzJjYWYyZjZlYjNkYWViZDUxMmIwMDgyMWRiZDNjOTExN2EwMGQ1YmVkNTEzYTc2ZTRmM2MmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.mUj23eJjpxF92UqUCmDulGkCZHWEp2ENyst9O2_nMgc)
> 
> The response.headers object is of type https://developer.mozilla.org/en-US/docs/Web/API/Headers so you i think you have to iterate over it to get the headers.
> 
> If it's not that, can you share some code that isn't working and the output of `tauri info` ?

My bad! By using `for (const p of myHeaders)`, the header can be obtained normally. Thank you!

**Comment by Arvmor** [comment]
Why not keeping the logic on Rust side, and only handle UI on frontend?

Easily can parse headers by using `ureq` or `reqwest`!
